### ISSUE TITLE
Can we get an user agent collector module ?

### ISSUE THREAD

**Issue by minanagehsalalma** [hak5/wifipineapple-modules#85]
Here's how it should be 
1- get the clients from other aps connected to our pineapple using deauth/karma
2- grab their user agent from their connection requests(apps and sevices ) or using a captive portal page to get it faster 
3- store their (device name - mac address - useragent - and the network name that they was connected to or their ssids prob requests ) in a db 
4- it would be great if we could automatically rank the vunrablite status of the devices from the db..

Thanks .... would do you think of this idea ?

**Comment by trashbo4t** [comment]
Not a bad idea. Do you know of any open databases that will return the vulnerability of a browser based on the UA?

If I were to design this module, I would use a captive portal that lets anyone through after they just land on it once. Store their metadata, and return.

**Comment by trashbo4t** [comment]
@Foxtrot if you assign this to me I can probably put something together.

@minanagehsalalma you can choose the module name :)

**Comment by minanagehsalalma** [comment]
> you can choose the module name :)

It's  my pleasure.. what about UAC stands for user agent collector 😅?

**Comment by minanagehsalalma** [comment]
> I would use a captive portal that lets anyone through after they just land on it once

Maybe not wait for them to land on a page but catch apps connection requests if possible to get the metadata even faster.

**Comment by trashbo4t** [comment]
@minanagehsalalma You would have to wait for a request made over HTTP, which would take longer no?

If you drop them into a "passive" captive portal they would hit your HTTP page, then you can grab the credentials and redirect them to wherever they were going.

**Comment by minanagehsalalma** [comment]
> You would have to wait for a request made over HTTP, which would take longer no?

Yup you are right ... But I am talking about in case of karma attack where the phone screen shouldn't be necessary turned on... So the apps in the background will do the trick I think.

**Comment by trashbo4t** [comment]
Ah I see, maybe it would be best to implement both, and neither if they are already captured.

**Comment by minanagehsalalma** [comment]
> best to implement both, and neither if they are already captured.

I agree it can't be better.

**Comment by Foxtrot** [comment]
Bear in mind that if you're trying to figure out client OS / Device type and not just strictly user agents, you can use more than HTTP requests.

**Comment by minanagehsalalma** [comment]
> not just strictly user agents

I think most of the time user agent contains the client os but maybe not it's version. Incase of mobile device it's either Android or iOS and you can tell that from just the model name.

**Comment by trashbo4t** [comment]
I think the easiest way to do this would be a captive portal. The second they connect to your device you can serve a blank page, snag the headers, then close the portal/authenticate the client.

**Comment by minanagehsalalma** [comment]
> I think the easiest way to do this would be a captive portal. The second they connect to your device you can serve a blank page, snag the headers, then close the portal/authenticate the client.

Yup that's the easiest way.

Although i don't think that capturing that header from apps requests is hard 
Just like [fluxion webserver tab](https://user-images.githubusercontent.com/1464758/68094837-35d1d800-fe59-11e9-8c7d-c99f1d293939.png)

**Comment by trashbo4t** [comment]
@minanagehsalalma  heres a small update on what I have working so far.

![](https://i.imgur.com/8G4RDHg.png)

I have the module drop a portal that authorizes users right away, while grabbing some profile information.

**Comment by minanagehsalalma** [comment]
> @minanagehsalalma heres a small update on what I have working so far.
> 
> ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> 
> I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> 
> I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.

great job mate this looks good so far.

**Comment by trashbo4t** [comment]
> > @minanagehsalalma heres a small update on what I have working so far.
> > ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> > I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> > I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.
> 
> great job mate this looks good so far.

Can you make a checklist on the original post for objectives like 

 - Capture HTTP Headers.
 - Capture MAC address.
 - Make Request to blah.com

**Comment by minanagehsalalma** [comment]
> > > @minanagehsalalma heres a small update on what I have working so far.
> > > ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> > > I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> > > I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.
> > 
> > 
> > great job mate this looks good so far.
> 
> Can you make a checklist on the original post for objectives like
> 
> * Capture HTTP Headers.
> * Capture MAC address.
> * Make Request to blah.com
> 
> That way I can go through each and check off points

Hi first where is the device model in the picture you uploaded just like this 
![2020-01-12 23 46 28](https://user-images.githubusercontent.com/20546638/72226107-ecf37980-3595-11ea-9a8c-14a322757d77.png)

**Comment by minanagehsalalma** [comment]
And about the objectives should i include the karma attack part ?

**Comment by trashbo4t** [comment]
> > > > @minanagehsalalma heres a small update on what I have working so far.
> > > > ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> > > > I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> > > > I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.
> > > 
> > > 
> > > great job mate this looks good so far.
> > 
> > 
> > Can you make a checklist on the original post for objectives like
> > 
> > * Capture HTTP Headers.
> > * Capture MAC address.
> > * Make Request to blah.com
> > 
> > That way I can go through each and check off points
> 
> Hi first where is the device model in the picture you uploaded just like this
> ![2020-01-12 23 46 28](https://user-images.githubusercontent.com/20546638/72226107-ecf37980-3595-11ea-9a8c-14a322757d77.png)

Here -> device is an Ipad
![image](https://user-images.githubusercontent.com/7047791/72226401-cc5bfd00-355e-11ea-9a82-f255a19196cb.png)

**Comment by minanagehsalalma** [comment]
> > > > > @minanagehsalalma heres a small update on what I have working so far.
> > > > > ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> > > > > I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> > > > > I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.
> > > > 
> > > > 
> > > > great job mate this looks good so far.
> > > 
> > > 
> > > Can you make a checklist on the original post for objectives like
> > > 
> > > * Capture HTTP Headers.
> > > * Capture MAC address.
> > > * Make Request to blah.com
> > > 
> > > That way I can go through each and check off points
> > 
> > 
> > Hi first where is the device model in the picture you uploaded just like this
> > ![2020-01-12 23 46 28](https://user-images.githubusercontent.com/20546638/72226107-ecf37980-3595-11ea-9a8c-14a322757d77.png)
> 
> Here -> device is an Ipad
> ![image](https://user-images.githubusercontent.com/7047791/72226401-cc5bfd00-355e-11ea-9a82-f255a19196cb.png)

Yup but shouldn't there be a model number just like the pic ?

**Comment by trashbo4t** [comment]
> And about the objectives should i include the karma attack part ?

I would leave the karma attack out of this module. 

This module is more like a Second-Stage attack.
Getting the user onto your network is the First-Stage.

I was thinking more like adding "Controls".

Like this:
![image](https://user-images.githubusercontent.com/7047791/72226427-12b15c00-355f-11ea-8313-e115fe37e8f3.png)


For example, alongside the captive portal we could have  a URL sniffer that pulls out profiling information. 

So the controls column reads like:

Captive Portal [On]
URLSniffer [On]
DNS Sniffer [On}
etc.
etc.

**Comment by trashbo4t** [comment]
> > > > > > @minanagehsalalma heres a small update on what I have working so far.
> > > > > > ![image](https://user-images.githubusercontent.com/7047791/72225293-de836e80-3551-11ea-9674-7db67d144668.png)
> > > > > > I have the module drop a portal that authorizes users right away, while grabbing some profile information.
> > > > > > I still have a lot of editing to do to make sure formatting looks good, as well as add some options like downloading these captures, etc. etc.
> > > > > 
> > > > > 
> > > > > great job mate this looks good so far.
> > > > 
> > > > 
> > > > Can you make a checklist on the original post for objectives like
> > > > 
> > > > * Capture HTTP Headers.
> > > > * Capture MAC address.
> > > > * Make Request to blah.com
> > > > 
> > > > That way I can go through each and check off points
> > > 
> > > 
> > > Hi first where is the device model in the picture you uploaded just like this
> > > ![2020-01-12 23 46 28](https://user-images.githubusercontent.com/20546638/72226107-ecf37980-3595-11ea-9a8c-14a322757d77.png)
> > 
> > 
> > Here -> device is an Ipad
> > ![image](https://user-images.githubusercontent.com/7047791/72226401-cc5bfd00-355e-11ea-9a82-f255a19196cb.png)
> 
> Yup but shouldn't there be a model number just like the pic ?

No idea, these are the Headers from the HTTP request.

**Comment by minanagehsalalma** [comment]
> could have a URL sniffer that pulls out profiling information.

What do you mean ? What will a url sniffer do in case of https requests ? 

> No idea, these are the Headers from the HTTP request.

Hmm.. can you try this site and check if it gives you the same results  !  
https://www.whatismybrowser.com/detect/what-is-my-user-agent

Also check the original post i have added the check list ... what do you think?

**Comment by trashbo4t** [comment]
> > could have a URL sniffer that pulls out profiling information.
> 
> Other than the useragent ?
> 
> > No idea, these are the Headers from the HTTP request.
> 
> Hmm.. can you try this site and check if it gives you the same results !
> https://www.whatismybrowser.com/detect/what-is-my-user-agent

I will just assume for now those Headers are correct. Perhaps Apple does not give away that info?

> Also check the original post i have added the check list ... what do you think?

Looking good so far

**Comment by minanagehsalalma** [comment]
> I will just assume for now those Headers are correct. Perhaps Apple does not give away that info?

Yup i checked  it and it doesn't 
Ref: https://deviceatlas.com/blog/mobile-browser-user-agent-strings

**Comment by minanagehsalalma** [comment]
> Edit : it sometimes does and sometimes doesn't
> Ref: https://51degrees.com/blog/detect-apple-iphone-model-numbers-and-user-agents

@Foxtrot what do you think the ios model number fix is ?

**Comment by Foxtrot** [comment]
Not sure what you mean. If the client browser isn't sending the model number in the User Agent string, then it simply is not there.

It's entirely up to the browser what is and isn't sent.

**Comment by minanagehsalalma** [comment]
> Not sure what you mean. If the client browser isn't sending the model number in the User Agent string, then it simply is not there.

Yup but i meant a way around it 
In the ref link the author said 
"In select cases, most commonly when requested through a web application such as Facebook, Snapchat or Instagram, we are treated with a device identifier."

So would connection requests from such applications contain it right ?

In this article there is much better solutions
https://51degrees.com/blog/device-detection-for-apple-iphone-and-ipad

Like using JavaScript to determine the phone model via it's screen height, width and pixel density. 

@trashbo4t should i add this to the check list ?

**Comment by minanagehsalalma** [comment]
@trashbo4t  does this gives you any additional info ? More than just the useragent for ios?
https://51degrees.com/resources/user-agent-tester

**Comment by minanagehsalalma** [comment]
This seems better and stable solution  [WURFL script from a stack over flow answer](https://stackoverflow.com/a/28057831/9889643) 
Just add this to the captive portal page 
`<script type='text/javascript' src=“//wurfl.io/wurfl.js"></script>`

And of course we can download it to get it working offline Then console.log(WURFL);

Or maybe this [iDevice.js](https://dieulot.fr/idevice) although  i think wurfl is better.

**Comment by minanagehsalalma** [comment]
@Foxtrot  do you know how did they catch such information ? 
> ```yaml
> user_agent: 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E216 [FBAN/FBIOS;FBAV/170.0.0.60.91;FBBV/105964764;FBDV/iPhone10,1;FBMD/iPhone;FBSN/iOS;FBSV/11.3;FBSS/2;FBCR/Sprint;FBID/phone;FBLC/en_US;FBOP/5;FBRV/106631002]'
> ```

Then they sort it like this 
![IMG_20200114_140923](https://user-images.githubusercontent.com/20546638/72343433-87a7a180-36d7-11ea-8f64-db8116b81036.png)

https://github.com/matomo-org/device-detector

https://github.com/matomo-org/device-detector/blob/master/regexes/device/mobiles.yml#L496

**Comment by minanagehsalalma** [comment]
@trashbo4t 
@Foxtrot 
![Screenshot_2020-01-15-16-25-56](https://user-images.githubusercontent.com/20546638/72441823-fd813b00-37b3-11ea-9eaf-9081061d7738.png)

Ehh... Were they watching the thread or what !!
https://www.zdnet.com/article/google-to-phase-out-user-agent-strings-in-chrome/

**Comment by minanagehsalalma** [comment]
At least it still works with fb or Ig... If we get them to open the captive portal page with them...
![IMG_20200115_164439](https://user-images.githubusercontent.com/20546638/72444067-0a079280-37b8-11ea-91f8-f6498e00d1b9.png)

**Comment by minanagehsalalma** [comment]
@trashbo4t  what do you think of the new one https://github.com/hak5/wifipineapple-modules/issues/88 it's has a bit in  common with this !

**Comment by trashbo4t** [comment]
@minanagehsalalma 

Perhaps a step by step implementation of your attack might be more readable.

Also, loved the google chrome article haha

**Comment by minanagehsalalma** [comment]
> loved the google chrome article haha
@trashbo4t 
Yup although that's gonna Impedes us a little bit ... It still a good thing .


> Perhaps a step by step implementation of your attack might be more readable.

Just check this ... It's the very same 
https://github.com/dxa4481/WPA2-HalfHandshake-Crack

Half handshake capturing (from failed connection by answering saved networks prob requests )and cracking them to find a password from a weak one that can be used later to get the victim auto connected to our fake ap  when there is no available saved open networks on his device.

**Comment by minanagehsalalma** [comment]
Or in numbered steps from another post 
1-capture the prob requests

2- launch 2 version of the ssids one open and one secure if it connects to the open one put a red check mark on it (in the list of the probed networks ) and if connects to the secure one capture the handshake and put a green check mark on it (in the same list )

3-after capturing a Good number of handshakes then start brute forcing

4- when it cracks a weak one.. broadcast it to get the victims connected

This should be a work around for karma attack when the targeted device doesn't have saved open networks.

**Comment by minanagehsalalma** [comment]
@trashbo4t 
I was thinking what about if we add a port scan to this ? 
So we know which ports are open on the device.. to identify the vunrablites better !

**Comment by trashbo4t** [comment]
> @trashbo4t
> I was thinking what about if we add a port scan to this ?
> So we know which ports are open on the device.. to identify the vunrablites better !

While I do think thats a good Idea, Im also keen on the idea of keeping modules "modular". 

As in separate functionality with a common means of communication. Something like microservices.

An nmap module already exists to do port scanning, so if that module does a port scan, users can collect that information and store it in their "Notes" with an association to a mac address.

Any other module that needs to know about open ports can run the scanning module, or reference their "Notes", and use that collected information.

**Comment by minanagehsalalma** [comment]
> > @trashbo4t
> > I was thinking what about if we add a port scan to this ?
> > So we know which ports are open on the device.. to identify the vunrablites better !
> 
> While I do think thats a good Idea, Im also keen on the idea of keeping modules "modular".
> 
> As in separate functionality with a common means of communication. Something like microservices.
> 
> An nmap module already exists to do port scanning, so if that module does a port scan, users can collect that information and store it in their "Notes" with an association to a mac address.
> 
> Any other module that needs to know about open ports can run the scanning module, or reference their "Notes", and use that collected information.
@trashbo4t 
Yup you are right

**Comment by minanagehsalalma** [comment]
But then there should be a merged module that can do all of this in a one run with the title of "clients info collector"

**Comment by trashbo4t** [comment]
> But then there should be a merged module that can do all of this in a one run with the title of "clients info collector"

What do you mean by merged?

**Comment by minanagehsalalma** [comment]
> What do you mean by merged?

All in one ..... Or they can just run the nmap module after this one.

**Comment by Foxtrot** [comment]
Modules can interact with other modules using the API, FYI.

```
$scope.func = (function() {
    $api.request({
         module: 'Recon',
         action: 'StartScan',
    }, function(response) {
        // etc...
    })
});
```

**Comment by minanagehsalalma** [comment]
> Modules can interact with other modules using the API, FYI.

Great we can add it to the description "to gather more info about the clients run the nmap module using the api"

**Comment by minanagehsalalma** [comment]
@trashbo4t 
https://github.com/hak5/wifipineapple-modules/issues/89

What do you think ? This the most somewhat new one i could come up with (currently  ;)
